dm/master/scheduler/scheduler.go

```diff
@@ -351,6 +353,66 @@ func (s *Scheduler) GetSourceCfgByID(source string) *config.SourceConfig {
 	return &clone
 }
 
+// TransferSource unbinds the source and binds it to a free worker. If fails halfway, the old worker should try recover
+func (s *Scheduler) TransferSource(source, worker string) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if !s.started {
+		return terror.ErrSchedulerNotStarted.Generate()
+	}
+
+	// 1. check existence or no need
+	if _, ok := s.sourceCfgs[source]; !ok {
+		return terror.ErrSchedulerSourceCfgNotExist.Generate(source)
+	}
+	w, ok := s.workers[worker]
+	if !ok {
+		return terror.ErrSchedulerWorkerNotExist.Generate(worker)
+	}
+	oldWorker, hasOldWorker := s.bounds[source]
+	if hasOldWorker && oldWorker.BaseInfo().Name == worker {
+		return nil
+	}
+
+	// 2. check new worker is free
+	stage := w.Stage()
+	if stage != WorkerFree {
+		return terror.ErrSchedulerWorkerInvalidTrans.Generate(worker, stage, WorkerBound)
+	}
+
+	// 3. if no old worker, bound it directly
+	if !hasOldWorker {
+		s.logger.Warn("in transfer source, found a free worker and not bound source, which should not happened",
+			zap.String("source", source),
+			zap.String("worker", worker))
+		err := s.boundSourceToWorker(source, w)
+		if err == nil {
+			delete(s.unbounds, source)
+		}
+		return err
+	}
+
+	// 4. replace the source bound
+	failpoint.Inject("failToReplaceSourceBound", func(_ failpoint.Value) {
+		failpoint.Return(errors.New("failToPutSourceBound"))
+	})
+	_, err := ha.ReplaceSourceBound(s.etcdCli, source, oldWorker.BaseInfo().Name, worker)
```

Comment:
Is there any problem when the last worker doesn't stop its own source timely?

Reply:
Delete source bound can't handle it? If so, we may have the same problem when there is a network isolation between worker and master.

Reply:
https://github.com/pingcap/dm/pull/1492/commits/30f87fbb3f182f6c1eb354a8f211d8a586ae2c17